File: stacks_and_queues/sliding_window_maximum.py
from collections import deque

def maxSlidingWindow(nums: List[int], k: int) -> List[int]:
    ans = []
    deque = deque()

    for i in range(len(nums)):
        # Maintain monotonically decreasing deque
        # All elements in the deque smaller than the current one
        # have no chance of being the maximum so get rid of them
        while deque and nums[i] > nums[deque[-1]]:
            deque.pop()
        
        deque.append(i)

        # deque[0] is the index of the maximum element
        # If deque[0] + k == i then it is outside the window
        if deque[0] + k == i:
            deque.popleft()
        
        # Only add to the answer once our window has reached size k
        if i >= k - 1:
            ans.append(nums[deque[0]])
    
    return ans

File: stacks_and_queues/sliding_window_maximum.js
var maxSlidingWindow = function(nums, k)
{
    let ans = [];
    let queue = [];

    for (let i = 0; i < nums.length; i++)
    {
        // Maintain monotonically decreasing queue
        // All elements in the queue smaller than the current one have no
        // chance of being the maximum so get rid of them
        while (queue.length && nums[i] > nums[queue[queue.length - 1]])
        {
            queue.pop();
        }

        queue.push(i);

        // queue[0] is the index of the maximum element
        // if queue[0] + k == i then it is outside the window
        if (queue[0] + k == i)
        {
            queue.shift();
        }

        // Only add to answer once our window has reached size k
        if (i >= k - 1)
        {
            ans.push(nums[queue[0]]);
        }
    }

    return ans;
};